brion/plugin/morphologyZeroEQ.cpp

```diff
@@ -0,0 +1,229 @@
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "morphologyZeroEQ.h"
+
+#include "../constants.h"
+
+#include <lunchbox/pluginRegisterer.h>
+#include <lunchbox/threadPool.h>
+#include <zeroeq/client.h>
+#include <zeroeq/uri.h>
+
+#include <mutex>
+
+namespace brion
+{
+namespace plugin
+{
+namespace
+{
+lunchbox::PluginRegisterer<MorphologyZeroEQ> registerer;
+}
+
+const std::string SERVER_SESSION("morphologyServer");
+namespace
+{
+lunchbox::ThreadPool& _getWorkers()
+{
+    static lunchbox::ThreadPool workers;
+    return workers;
+}
+}
+
+class MorphologyZeroEQ::Client // adds thread-safety to zeroeq::Client
+{
+public:
+    Client()
+        : _client(getenv(zeroeq::ENV_REP_SESSION.c_str())
+                      ? zeroeq::DEFAULT_SESSION
+                      : SERVER_SESSION)
+    {
+    }
+
+    Client(const zeroeq::URI& uri)
+        : _client({uri})
+    {
+    }
+
+    bool request(const uint128_t& r, const void* data, size_t size,
+                 const zeroeq::ReplyFunc& func)
+    {
+        std::lock_guard<std::mutex> lock(_mutex);
+        return _client.request(r, data, size, func);
+    }
+
+    bool receive()
+    {
+        std::lock_guard<std::mutex> lock(_mutex);
+        return _client.receive(0);
+    }
+
+    void lock() { _mutex.lock(); }
+    void unlock() { _mutex.unlock(); }
+private:
+    std::mutex _mutex;
+    zeroeq::Client _client;
+};
+
+MorphologyZeroEQ::MorphologyZeroEQ(const MorphologyInitData& initData)
+    : MorphologyPlugin(initData)
+    , _client(_getClient())
+{
+    ClientPtr client = _client; // keep ref for thread-safety
+    const std::string path = initData.getURI().getPath();
+    const auto handler = [&](const uint128_t& id, const void* data,
+                             const size_t size) {
+        if (id == 0)
+            LBWARN << "Server could not load morphology" << std::endl;
+        if (data && size)
+        {
+            _client->unlock();
+            _data.fromBinary(data, size);
+            _client->lock();
+        }
+        _client.reset();
+    };
+
+    if (!client->request(ZEROEQ_GET_MORPHOLOGY, path.data(), path.size(),
+                         handler))
+    {
+        LBTHROW(std::runtime_error("Failed to request morphology data"));
+    }
+
+    _loader = _getWorkers().post([&] {
+        ClientPtr client_ = _client; // keep ref for thread-safety
+        while (_client)
+            client_->receive();
```

Comment:
Nice horse you've got there. 

Having workers for the fromBinary means you need to memcpy the data out of the zmq_msg which will certainly be slower. We could extend the ZeroEQ API to allow ownership transfer of the zmq messages, but at what cost to resolve one minor itch here? After all, these threads only spin as long as data is loading.

I've done a quick test (e06a167) having only one thread with blocking receive, and thats slower in all cases (17k vs 12k/s locally, 3k vs .8k remotely from Lugano).

BTW: zmq_poll seems to spin too. At least the morphologyServer (with a blocking receive) hogs a CPU too.